loader-tests/index.ts

```diff
@@ -0,0 +1,115 @@
+import './dom-parser.js'
+
+import {
+  clearPreview,
+  createDownloadTask,
+  createTextResponse,
+  enableTestTime,
+  type EnvironmentAndStore,
+  getLoaderForText,
+  getTestEnvironment,
+  loaders,
+  type PreviewCandidate,
+  previewCandidates,
+  setBaseRoute,
+  setPreviewUrl,
+  setRequestMethod,
+  setupEnvironment,
+  userId
+} from '@slowreader/core'
+import { keepMount } from 'nanostores'
+import { readFile } from 'node:fs/promises'
+import { setTimeout } from 'node:timers/promises'
+
+import { LoaderNotFoundError, UnsupportedFileExtError } from './errors.js'
+import { isString, isSupportedExt, logger, resolvePath } from './utils.js'
+
+interface Feed {
+  htmlUrl: string
+  title: string
+  url: string
+}
+
+const timerDurationInSec = 10
+
+async function parseFeedsFromFile(path: string): Promise<void> {
+  try {
+    if (!isSupportedExt(path)) {
+      throw new UnsupportedFileExtError(path)
+    }
+    let buffer = await readFile(resolvePath(path))
+    let text = createTextResponse(buffer.toString('utf-8'))
+    let feeds: Feed[] = [...text.parse().querySelectorAll('[type="rss"]')]
+      .filter(feed => isString(feed.getAttribute('xmlUrl')))
+      .map(f => ({
+        htmlUrl: f.getAttribute('htmlUrl')!,
+        title: f.getAttribute('title') || '',
+        url: f.getAttribute('xmlUrl')!
+      }))
+    await Promise.all([...feeds.map(feed => fetchAndParsePosts(feed))])
+    await Promise.all([...feeds.map(feed => findRSSfromHome(feed))])
+  } catch (e) {
+    if (e instanceof Error && 'code' in e && e.code === 'ENOENT') {
+      logger.err(`File not found on path: "${path}". \n`)
+    } else if (e instanceof Error) {
+      logger.err(`${e.message}\n`)
+    } else if (typeof e === 'string') {
+      logger.err(e)
+    }
+  }
+}
+
+async function fetchAndParsePosts(feed: Feed): Promise<void> {
+  try {
+    let task = createDownloadTask()
+    let textResponse = await task.text(feed.url)
+    let candidate: false | PreviewCandidate = getLoaderForText(textResponse)
+    if (!candidate) {
+      throw new LoaderNotFoundError(feed.title)
+    }
+    let loader = loaders[candidate.loader]
+    let { list: posts } = loader.getPosts(task, feed.url, textResponse).get()
+    logger.succ(`[Feed] - ${feed.title} - ${feed.url}`)
+    logger.succ(`• Found ${posts.length} post(s)\n`)
+  } catch (e) {
+    if (e instanceof Error) {
+      logger.err(`${e.message}\n`)
+    } else if (typeof e === 'string') {
+      logger.err(`${e}\n`)
+    }
+  }
+}
+
+async function findRSSfromHome(feed: Feed): Promise<void> {
+  try {
+    keepMount(previewCandidates)
+    await setPreviewUrl(feed.htmlUrl)
+    await setTimeout(timerDurationInSec * 1000)
+    if (previewCandidates.get().length === 0) {
```

Comment:
Timeout code can be a little more complicated
https://advancedweb.hu/how-to-add-timeout-to-a-promise-in-javascript/

Reply:
@ai sure, but check this out... if we just leave the fn like this:
```javascript 
function waitForTrue(store) {
  return new Promise(resolve => {
    let unbind = store.listen(value => {
      if (value === true) {
        unbind()
        resolve()
      }
    })
  }
}
```

and then use it like you said : ```javascript
waitForTrue(previewCandidatesLoading),
```

It will resolve immediately when *isLoading* is set from `false` to `true`, to indicate the loading process, not when the promise actually resolved.

So, the logic to wait for resolvement would have to be something like this:
```typescript
export function waitForStoreResolve(
  store: ReadableAtom<boolean>
): Promise<boolean> {
  return new Promise(resolve => {
    let loadingComparison = true
    let unbind = store.listen(value => {
      if (value && loadingComparison) {
        /**
         * value = true actually is an indicator of "loading in progress"
         * value changed from 'idle' to 'isLoading'
         * since
         *  */
        loadingComparison = false
      }

      if (!value && !loadingComparison) {
        /**
         * here it actually resolved, changed from 'isLoading' to 'done'
         * value = true -> value = false
         * it was loading and now is not anymore
         * */
        unbind()
        resolve(value)
      }
    })
  })
}
```

if this is not too hacky - I can include it.
I felt like it was too much so I just repeated what we did in the tests with the "stupid" timer.

Reply:
Yes, let’s go this direction. I will try to clean up code later.